s3transfer/manager.py

```diff
@@ -378,11 +389,106 @@ def _register_handlers(self):
+        :param cancel: If True, calls TransferFuture.cancel() for
+            all in-progress in transfers. This is useful if you want the
+            shutdown to happen quicker.
+        """
+        if cancel:
+            # Cancel all in-flight transfers if requested, before waiting
+            # for them to complete.
+            self._canceler.cancel()
+        try:
+            # Wait until there are no more in-progress transfers. This is
+            # wrapped in a try statement because this can be interrupted
+            # with a KeyboardInterrupt that needs to be caught.
+            self._canceler.wait()
+        finally:
+            # If not errors were raised in the try block, the cancel should
+            # have no coordinators it needs to run cancel on. If there was
+            # an error raised in the try statement we want to cancel all of
+            # the inflight transfers before shutting down to speed that
+            # process up.
+            self._canceler.cancel()
+            # Shutdown all of the executors.
+            self._submission_executor.shutdown()
+            self._request_executor.shutdown()
+            self._io_executor.shutdown()
+
+
+class TransferCoordinatorCanceler(object):
+    def __init__(self):
+        """Abstraction to control the cancelation of all transfers
+
+        This abstraction allows the manager to wait for inprogress transfers
+        to complete and cancel all inprogress transfers.
+        """
+        self._lock = threading.Lock()
+        self._active_transfer_coordinators = set()
+
+    def add_transfer_coordinator(self, transfer_coordinator):
+        """Adds a transfer coordinator of a transfer to be canceled if needed
+
+        :type transfer_coordinator: s3transfer.futures.TransferCoordinator
+        :param transfer_coordinator: The transfer coordinator for the
+            particular transfer
+        """
+        with self._lock:
+            self._active_transfer_coordinators.add(transfer_coordinator)
+
+    def remove_transfer_coordinator(self, transfer_coordinator):
+        """Remove a transfer coordinator from cancelation consideration
+
+        Typically, this method is invoked by the transfer coordinator itself
+        to remove its self when it completes its transfer.
+
+        :type transfer_coordinator: s3transfer.futures.TransferCoordinator
+        :param transfer_coordinator: The transfer coordinator for the
+            particular transfer
+        """
+        with self._lock:
+            self._active_transfer_coordinators.remove(transfer_coordinator)
+
+    def cancel(self):
+        """Cancels all inprogress transfers
+
+        This cancels the inprogress transfers by calling cancel() on all
+        tracked transfer coordinators.
+        """
+        with self._lock:
+            for transfer_coordinator in self._active_transfer_coordinators:
+                transfer_coordinator.cancel()
+
+    def wait(self, sleep_intervals=0.2):
```

Comment:
Yeah, this is something it took me a long time to figure out myself in a similar application, since the behavior doesn't seem to be documented anywhere.

I'm hoping this will become my go-to library for interfacing with S3 for bulk data transfers. Good to see it coming together. Please consider tracking performance and memory use in the test suite too, to avoid regressions. I've seen some severe transport performance regressions make it into releases in even the highest profile projects like requests and node.